### ISSUE TITLE
Generate opcode metadata from bytecodes.c instead of opcode.py

### ISSUE THREAD

**Issue by iritkatriel** [python/cpython#105481]
We would ideally have bytecodes.c as the single source of truth about opcodes. So opcode.py and the code generated from it should be replaced by alternatives from the cases_generator, if we can.

### Linked PRs
* gh-105482
* gh-105506
* gh-105788
* gh-105791
* gh-105865
* gh-105913
* gh-105950
* gh-106673
* gh-106688
* gh-106758
* gh-106913
* gh-107276
* gh-107339
* gh-107534
* gh-107540
* gh-107561
* gh-107598
* gh-107866
* gh-107942
* gh-107971
* gh-108025
* gh-108079
* gh-108080
* gh-108367

**Comment by iritkatriel** [comment]
@gvanrossum The old macros HAS_ARG and HAS_CONST are exposed in Include/opcode.h (and have been for a long time). Does this mean that before we can remove them, we need to expose the opcode_metadata.h file there as well? We should also mark all of this as unstable API (how do we do that?)

**Comment by gvanrossum** [comment]
Neither `HAS_ARG` nor `HAS_CONST` is mentioned in the docs at all. I am of the opinion that in this case this implies they are not public, and anyone using them should not be surprised if they disappear. (It would be more of an issue if they turned into lies.)

opcode.h is mentioned only once in the docs, as the source of truth for dis.py. (It is also mentioned twice in Misc/NEWS.d/, but that's really just a changelog.

I do think we ought to provide similar functionality to debuggers.

Info about declaring unstable APIs comes from PEP 649, which links to https://devguide.python.org/developer-workflow/c-api/#c-api. It looks like you must use the `PyUnstable_` prefix (e.g. `PyUnstable_BytecodeHasArg`), and it ought to be in a .h file under Include/cpython/.

**Comment by iritkatriel** [comment]
> I do think we ought to provide similar functionality to debuggers.

Do you mean to expose the stuff in opcode_metadata.h file through some unstable c api?

**Comment by gvanrossum** [comment]
Yeah, to the extent that it’s useful, in function form (so the data structure is not public).

Or we can wait until someone asks.

**Comment by iritkatriel** [comment]
There is a usage of HAS_ARG in Modules/_opcode.c.    We could remove it (send oparg -1 if None was given and do this check inside Python/ code). Otherwise we need to move opcode_metadata.h from Python/ to Include/internal.

**Comment by vstinner** [comment]
@encukou wrote "Unstable API needs documentation and tests."

I reopen this issue and close gh-107149.

**Comment by iritkatriel** [comment]
I'm not sure it was the right move to rename `_PyUnstable_GetUnaryIntrinsicName` to `PyUnstable_GetUnaryIntrinsicName`, etc.

I created them as private because I did not intend to document them as a public API (we need a reason to do that). I put Unstable in the name because it's one of those things that can change all the time. If we don't want the `_PyUnstable_` prefix then let's remove the `Unstable` and leave the `_` until we have a reason/decision to make these part of the public C API.
